mlir/lib/Dialect/Rock/Transforms/GemmToGridwise.cpp

```diff
@@ -64,6 +67,51 @@ struct AttentionRewritePattern : public OpConversionPattern<AttentionOp> {
   LogicalResult matchAndRewrite(AttentionOp op, AttentionOpAdaptor adaptor,
                                 ConversionPatternRewriter &b) const override;
 };
+
+static Type getSmallestType(Type type1, Type type2) {
+  return (type1.getIntOrFloatBitWidth() > type2.getIntOrFloatBitWidth())
+             ? type2
+             : type1;
+}
+
+static Type deduceAccumulatorElementType(Type elementTypeA, Type elementTypeB,
+                                         Type elementTypeC,
+                                         OpBuilder &builder) {
+  // Determine the type used on VGPR to act as accumulator.
+  // f32: f32.
+  // f16, bf16: f32 to prevent overflow from happening.
+  // i16 : i16.
+  // fp8 (any combo) : f32.
+  // i8: i32, since we have an i32 output
+  auto type = getSmallestType(elementTypeA, elementTypeB);
+  if (type.isa<FloatType>() && type.getIntOrFloatBitWidth() < 32) {
+    return builder.getF32Type();
+  } else if (type.isInteger(8)) {
+    return builder.getI32Type();
+  }
+  return elementTypeC;
+}
+
+static auto getAccumulator(Value a, Value b, Value c, OpBuilder &builder,
+                           Location &loc) {
+  auto aElementType = a.getType().cast<MemRefType>().getElementType();
+  auto bElementType = b.getType().cast<MemRefType>().getElementType();
+  auto cElementType = c.getType().cast<MemRefType>().getElementType();
+
+  auto accumulatorElementType = deduceAccumulatorElementType(
+      aElementType, bElementType, cElementType, builder);
+
+  if (accumulatorElementType != cElementType) {
+    auto accumulatorShape = c.getType().cast<MemRefType>().getShape();
+    auto accumulatorType =
+        MemRefType::get(accumulatorShape, accumulatorElementType);
+    Value accumulator = builder.create<memref::AllocOp>(loc, accumulatorType);
+    return std::make_pair(accumulator, true);
+  } else {
+    return std::make_pair(c, false);
```

Comment:
In theory, you could swap this flag for `accumulator == c` but that's a design decision

Reply:
Hmm, I decided to return just a `Value` from `getAccumulator` (in stead of a pair). I moved the check `accumulator != c` down to `GemmRewritePattern::matchAndRewrite`.

I didn't expect the comparison operator of `Type` checking both shapes and element types. I mean it makes sense; just didn't expect it for some reason.